### PR TITLE
ARROW-17953: [Archery] Add archery docker info command

### PR DESCRIPTION
**Pull request by raulcd**
Some usage examples:

```
$ archery docker info debian-go-cgo --show command
Service debian-go-cgo docker compose config:
 - command: /bin/bash -c "
  /arrow/ci/scripts/go_build.sh /arrow &&
  /arrow/ci/scripts/go_test.sh /arrow"
```
or
```
$ archery docker info debian-go-cgo
Service debian-go-cgo docker compose config:
 - image: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo
 - build
  - context: .
  - dockerfile: ci/docker/debian-go-cgo.dockerfile
  - cache_from: ['${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo']
  - args
   - base: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}
 - shm_size: 2G
 - volumes: ['.:/arrow:delegated', '${DOCKER_VOLUME_PREFIX}debian-ccache:/ccache:delegated']
 - environment
  - ARROW_GO_TESTCGO: 1
 - command: /bin/bash -c "
  /arrow/ci/scripts/go_build.sh /arrow &&
  /arrow/ci/scripts/go_test.sh /arrow"
```
or
```
$ archery docker info ubuntu-cuda-cpp --show environment
Service ubuntu-cuda-cpp docker compose config:
 - environment
  - ARROW_BUILD_STATIC: OFF
  - ARROW_CUDA: ON
  - ARROW_GANDIVA: OFF
  - ARROW_GCS: OFF
  - ARROW_ORC: OFF
  - ARROW_S3: OFF
  - ARROW_SUBSTRAIT: OFF
  - ARROW_WITH_OPENTELEMETRY: OFF
  - CCACHE_COMPILERCHECK: content
  - CCACHE_COMPRESS: 1
  - CCACHE_COMPRESSLEVEL: 6
  - CCACHE_MAXSIZE: 1G
  - CCACHE_DIR: /ccache

```
_edited only the commands to reflect the change of the command --show_

### CODE REVIEW

**Review by assignUser — APPROVED** [review]
Cool idea! Looks good overall, As a minor thing I am not sure if I like the addition of `-` for structuring the output or if it just adds noise and only indentation would be cleaner?

File: dev/archery/archery/docker/cli.py [hunk 1]
```diff
@@ -289,3 +289,26 @@ def docker_compose_images(obj):
...
+    """Show docker-compose definition info for service_name.
+
+    SERVICE_NAME is the name of the docker service defined on
+    the docker-compose. Look at `archery docker images` output for names.
+    """
+    compose = obj['compose']
+    try:
+        service = compose.config.raw_config["services"][service_name]
+    except KeyError:
+        click.echo(f'Service name {service_name} could not be found')
```
Comment (pitrou) [inline dev/archery/archery/docker/cli.py:310]: This should behave more as an error, so for example:
```suggestion
        click.echo(f'Service name {service_name} could not be found', err=True)
        sys.exit(1)
```

File: dev/archery/archery/docker/cli.py [hunk 2]
```diff
@@ -289,3 +289,26 @@ def docker_compose_images(obj):
...
+    else:
+        click.echo(f'Service {service_name} docker compose config:')
```
Comment (pitrou) [inline dev/archery/archery/docker/cli.py:312]: Nit
```suggestion
        click.echo(f'Service {service_name} docker-compose config:')
```

File: dev/archery/archery/docker/core.py [hunk 3]
```diff
@@ -417,3 +417,19 @@ def _push(service):
...
+        output = []
+        for key, value in key_name.items():
+            if hasattr(value, 'items'):
+                temp_filters = filters
+                if key == filters or filters is None:
+                    output.append(f'{prefix}- {key}')
+                    # Keep showing this specific key
+                    # as parent matched filter
+                    temp_filters = None
+                output.extend(self.info(value, temp_filters, prefix + " "))
```
Comment (pitrou) [inline dev/archery/archery/docker/core.py:431]: Should probably use two spaces for better and more visible alignment?

File: dev/archery/archery/docker/cli.py [hunk 4]
```diff
@@ -289,3 +289,26 @@ def docker_compose_images(obj):
     click.echo('Available images:')
     for image in compose.images():
         click.echo(f' - {image}')
+
+
+@docker.command('info')
+@click.argument('service_name')
+@click.option('--only', '-o', required=False,
```
[the remark this answers is not in the document]
Comment (raulcd) [reply dev/archery/archery/docker/cli.py]: I've updated to `-s` or `--show`

File: dev/archery/archery/docker/core.py (hunk 3 above)
Comment (raulcd) [reply dev/archery/archery/docker/core.py]: good idea, with that I can remove the `-` and have a more visible alignment:
```
$ archery docker info debian-go-cgo 
Service debian-go-cgo docker-compose config:
  image: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo
  build
    context: .
    dockerfile: ci/docker/debian-go-cgo.dockerfile
    cache_from: ['${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}-cgo']
    args
      base: ${REPO}:${ARCH}-debian-${DEBIAN}-go-${GO}
  shm_size: 2G
  volumes: ['.:/arrow:delegated', '${DOCKER_VOLUME_PREFIX}debian-ccache:/ccache:delegated']
  environment
    ARROW_GO_TESTCGO: 1
  command: /bin/bash -c "
  /arrow/ci/scripts/go_build.sh /arrow &&
  /arrow/ci/scripts/go_test.sh /arrow"
```

**Comment by raulcd** [comment]
@pitrou @assignUser this should be ready to review again. Thanks for the feedback!

**Review by kou — APPROVED** [review]
+1

**Review by assignUser — APPROVED** [review]
Tested it locally, looks good :+1:

**Comment by pitrou** [comment]
I've noticed something:
```
$ archery docker info conda-cpp
Service conda-cpp docker-compose config:
  image: ${REPO}:${ARCH}-conda-cpp
  build
    context: .
    dockerfile: ci/docker/conda-cpp.dockerfile
    cache_from: ['${REPO}:${ARCH}-conda-cpp']
    args
      repo: ${REPO}
      arch: ${ARCH}
  shm_size: 2G
  ulimits
    core: ${ULIMIT_CORE}
  environment
    ARROW_BUILD_BENCHMARKS: ON
    ARROW_BUILD_EXAMPLES: ON
    ARROW_ENABLE_TIMING_TESTS: None
    ARROW_EXTRA_ERROR_CONTEXT: ON
    ARROW_MIMALLOC: ON
    ARROW_USE_LD_GOLD: ON
    ARROW_USE_PRECOMPILED_HEADERS: ON
    BUILD_DOCS_PYTHON: ON
    CCACHE_COMPILERCHECK: content
    CCACHE_COMPRESS: 1
    CCACHE_COMPRESSLEVEL: 6
    CCACHE_MAXSIZE: 1G
    CCACHE_DIR: /ccache
    AWS_ACCESS_KEY_ID: None
    AWS_SECRET_ACCESS_KEY: None
    SCCACHE_BUCKET: None
    SCCACHE_S3_KEY_PREFIX: ${SCCACHE_S3_KEY_PREFIX:-sccache}
  volumes: ['.:/arrow:delegated', '${DOCKER_VOLUME_PREFIX}conda-ccache:/ccache:delegated']
  command: [' /arrow/ci/scripts/cpp_build.sh /arrow /build && /arrow/ci/scripts/cpp_test.sh /arrow /build']
```

Printing `ARROW_ENABLE_TIMING_TESTS: None` is a bit unfortunate. What happens actually is that the value is inherited from the calling shell's environment. So perhaps print out `<inherited>` or something.

**Comment by assignUser** [comment]
+1 
```
Service conda-cpp docker-compose config:
  image: ${REPO}:${ARCH}-conda-cpp
  build
    context: .
    dockerfile: ci/docker/conda-cpp.dockerfile
    cache_from: ['${REPO}:${ARCH}-conda-cpp']
    args
      repo: ${REPO}
      arch: ${ARCH}
  shm_size: 2G
  ulimits
    core: ${ULIMIT_CORE}
  environment
    ARROW_BUILD_BENCHMARKS: ON
    ARROW_BUILD_EXAMPLES: ON
    ARROW_ENABLE_TIMING_TESTS: <inherited>
    ARROW_EXTRA_ERROR_CONTEXT: ON
    ARROW_MIMALLOC: ON
    ARROW_USE_LD_GOLD: ON
    ARROW_USE_PRECOMPILED_HEADERS: ON
    BUILD_DOCS_PYTHON: ON
    CCACHE_COMPILERCHECK: content
    CCACHE_COMPRESS: 1
    CCACHE_COMPRESSLEVEL: 6
    CCACHE_MAXSIZE: 1G
    CCACHE_DIR: /ccache
    AWS_ACCESS_KEY_ID: <inherited>
    AWS_SECRET_ACCESS_KEY: <inherited>
    SCCACHE_BUCKET: <inherited>
    SCCACHE_S3_KEY_PREFIX: ${SCCACHE_S3_KEY_PREFIX:-sccache}
  volumes: ['.:/arrow:delegated', '${DOCKER_VOLUME_PREFIX}conda-ccache:/ccache:delegated']
  command: [' /arrow/ci/scripts/cpp_build.sh /arrow /build && /arrow/ci/scripts/cpp_test.sh /arrow /build']
```

**Review by pitrou — APPROVED** [review]
Thanks @raulcd !

**Pull request merged**